whiteout/resolver.go

```diff
@@ -0,0 +1,94 @@
+package whiteout
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+
+	"github.com/quay/zlog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/indexer"
+)
+
+var (
+	_ indexer.Resolver = (*Resolver)(nil)
+)
+
+type Resolver struct{}
+
+func (r *Resolver) Resolve(ctx context.Context, ir *claircore.IndexReport, layers []*claircore.Layer) *claircore.IndexReport {
+	// Here we need to check if any of the packages
+	// found are moot due to whiteouts.
+	ls := newLayerSorter(layers)
+	finalPackages := map[string]*claircore.Package{}
+	finalEnvironments := map[string][]*claircore.Environment{}
+	for pkgID, pkg := range ir.Packages {
+		packageDeleted := false
+		// First env.IntroducedIn?
+		packageLayer := ir.Environments[pkgID][0].IntroducedIn.String()
+		for fileLayer, f := range ir.Files {
+			// Check if it's a whiteout file, if it applies to the package's
+			// file and if the layer the whiteout file came from came after.
+			if f.Kind == claircore.FileKindWhiteout && ls.isChildOf(fileLayer, packageLayer) && fileIsDeleted(pkg.Filepath, f.Path) {
+				packageDeleted = true
+				zlog.Debug(ctx).
+					Str("package name", pkg.Name).
+					Str("package file", pkg.Filepath).
+					Str("whiteout file", f.Path).
+					Msg("package determined to be deleted")
+			}
+		}
+		if !packageDeleted {
+			finalPackages[pkgID] = pkg
+			finalEnvironments[pkgID] = ir.Environments[pkgID]
+
+		}
+	}
+	ir.Packages = finalPackages
+	ir.Environments = finalEnvironments
+	return ir
+}
+
+// FileIsDeleted returns whether or not the filepath(fp) has been deleted
+// by the corresponding whiteoutPath. It follows the OCI spec for whiteouts:
+// https://github.com/opencontainers/image-spec/blob/main/layer.md#whiteouts
+func fileIsDeleted(fp, whiteoutPath string) bool {
+	var checkFile string
+	switch {
+	case filepath.Base(whiteoutPath) == ".wh..wh..opq":
+		// Special opaque case, "indicating that all siblings are hidden in the lower layer"
+		checkFile = filepath.Dir(whiteoutPath)
+	case strings.HasPrefix(filepath.Base(whiteoutPath), ".wh."):
+		origFileName := filepath.Base(whiteoutPath)[4:]
+		checkFile = filepath.Join(filepath.Dir(whiteoutPath), origFileName)
+	default:
+		return false
+	}
+
+	checkFileParts := strings.Split(checkFile, "/")
+	fpParts := strings.Split(fp, "/")
+	if len(checkFileParts) > len(fpParts) {
+		return false
+	}
+	for i, p := range checkFileParts {
+		if p != fpParts[i] {
+			return false
+		}
+	}
+	return true
+}
+
+type layerSorter map[string]int
+
+func newLayerSorter(layers []*claircore.Layer) layerSorter {
+	ls := layerSorter{}
```

Comment:
up to you since it's just a suggestion to the runtime and not guaranteed, but maybe initialize with the known length: `make(map[string]int, len(layers))`